Big puppet refactor - deployment scripts

binary1230 opened the pull request:
# DO NOT MERGE!!

This code is now ready to go, but, needs to be merged extremely carefully. Only Dom should merge it.
# summary
- install different facter facts (used by puppet to figure out how to classify things) for: environments, event_name, and vagrant yes/no
- minor overhaul of deploy scripts to deal with this
- introduce puppet Roles pattern which assigns a Role to this server which lets it include all ubersystem classes needed
- update hiera.yaml to be way more robust and deal with multiple hierarchy levels, described in https://github.com/magfest/production-config/pull/26

requires:
- https://github.com/magfest/ubersystem-puppet/pull/37
- https://github.com/magfest/production-config/pull/26

binary1230 committed:
horrbily broken WIP puppet refactor, REBASE START HERE

binary1230 committed:
working example of stuff

binary1230 committed:
working example of stuff, inheritance now

binary1230 committed:
psql support

binary1230 committed:
add in support for hiera() looks up in hiera files themselves
updated shortcuts
add in uber::nginx into rams_app_stack

binary1230 committed:
more wip

binary1230 committed:
even more wip!

binary1230 committed:
WIP, restructuring pretty much completed
now, getting it all to work again

binary1230 committed:
mostly working puppet setup once more
needs to be tested and fix dependency junk

binary1230 committed:
force python_setup to happen AFTER git clone of sideboard/plugins
dependency work is now complete, I think

binary1230 committed:
set this back to normal

binary1230 committed:
first start on reworking hiera, long way to go

binary1230 committed:
getting some deploy settings ready to go

binary1230 committed:
fabfile now puts custom per-node facts when you create a node

binary1230 committed:
vagrant we'll use port 443 internally and forward to 4443
recent refactorings should no longer require internal ubersystem
to keep the same port number internally and externally

binary1230 committed:
restructure all hiera data
- it's now in the right place, but totally wrong format
- next up we'll fix the formatting

binary1230 committed:
restructure all hiera data, part 2
- now it's potentialy working, but untested, config

binary1230 committed:
don't specify magfest-specific defaults

binary1230 committed:
update docs

binary1230 committed:
add some syntactical sugar to make puppet not complain

binary1230 committed:
fix control server script to correctly accept event_name as an arg

binary1230 committed:
Merge remote-tracking branch 'origin/big_puppet_refactor' into big_puppet_refactor

binary1230 committed:
fixes port forwarding
- experiment that didn't work. le sigh.

binary1230 committed:
add debug comments

binary1230 committed:
use correct db information

binary1230 committed:
use correct SSL port for vagrant

binary1230 commented on puppet/hiera/hiera.yaml:
```diff
@@ -1,13 +1,21 @@
+  - "vagrant-%{::is_vagrant}"
+  - "%{::environment}"
+  - "common"
+# :merge_behavior: deeper # used to need this, probably not anymore.
```
this is really the heart and crux of the refactor.  if you don't look at too much else, definitely pay attention to this.

it includes the files starting from "common" and works its way up, so that settings from below are overriden and added on to by files higher up in the hierarchy.

kitsuta replied:
OK, this actually answers some of the questions I was asking in the other PR.

But under this structure, if you put something in development.yaml, it's overridden by event-specific configs.

Does this mean that we'll need to treat each staging server as an event, instead of having it inherit the event config and overridding a few values for testing?

binary1230 replied:
So the way this works is, each server is, before puppet is run, assigned a set of 'facts' through a program called Facter.

In the case of let's say, stage.uber.magfest.org, the facts would be:
event_name='classic'
environment='development'
is_vagrant='false'
fqdn='stage.uber.magfest.org' (fully qualified domain name)

For prime.uber.magfest.org it would be
event_name='prime'
environment='production'
is_vagrant='false'
fqdn='prime.uber.magfest.org'

For a local vagrant instance, it would be:
event_name='[whatever you want it to be]'
environment='development'
is_vagrant='true'
fqdn='localhost.localdomain'

These facts are used to include certain files according to hiera.yaml.  Any settings set in both a lower file and a higher file is hidden by stuff in the higher up file:
```
"nodes/external/secret/%{::fqdn}"
"nodes/external/secret/event-%{::event_name}-%{::environment}"
"nodes/external/secret/event-%{::event_name}"
"nodes/external/secret/%{::environment}"
"nodes/external/secret/common"
"nodes/external/%{::fqdn}"
"nodes/vagrant-%{::event_name}-%{::is_vagrant}"
"nodes/vagrant-%{::is_vagrant}"
"nodes/event-%{::event_name}"
"nodes/%{::environment}"
"nodes/common"
"vagrant-%{::is_vagrant}"
"%{::environment}"
"common"
```

(note: for this example I added a few entries not present cause I haven't gotten to them yet)

For stage.uber.magfest.org it looks like this:
```
"nodes/external/secret/stage.uber.magfest.org.yaml"
"nodes/external/secret/event-classic-development.yaml"
"nodes/external/secret/event-classic.yaml"
"nodes/external/secret/development.yaml"
"nodes/external/secret/common.yaml"
"nodes/external/stage.uber.magfest.org.yaml"
"nodes/vagrant-classic-0.yaml"
"nodes/vagrant-0.yaml"
"nodes/event-classic.yaml"
"nodes/development.yaml"
"nodes/common.yaml"
"vagrant-0.yaml"
"development.yaml"
"common.yaml"
```

Not all of those files exist (for instance, vagrant-0.yaml never exists and stage.uber,magfest.org,yaml) so here's a list of just the ones that do exist or aren't placeholders:
```
"nodes/external/secret/event-classic-development.yaml"
"nodes/external/secret/common.yaml"
"nodes/external/secret/development.yaml"
"nodes/event-classic.yaml"
"nodes/common.yaml"
"common.yaml"
```

Going node by node here for that stuff:
common.yaml - default settings we expect all events to use.  things like system settings, database access control settings, etc.  which core ubersystem plugins to use (uber, analytics, etc)
nodes/ -> everything in here is magfest-specific and only included by magfest.
nodes/common.yaml -> magfest-specific default settings.  things like email signatures, magfest shirt sizes, SSL certificates for magfest.org, organization_name='Magfest INC'.  basically anything in common between magprime, magclassic, magstock, rockage
nodes/event-classic.yaml -> the vast majority of magfest-specific settings. badge prices, dates, food prices, dept head checklists, any event-specific plugins to install etc
nodes/external/secret/common.yaml -> secret settings common to all magfest events.  for us, it's AWS keys.
nodes/external/secret/event-classic-development.yaml -> Stripe keys - testing versions only



For other events like Anthrocon, we'd expect them to supply a completely new nodes/ dir and plug in their specific settings there.

Anyway, I think everyone's going to find that these hierarchy settings make a ton of sense and make adding changes easier.  What's really cool is that, for example, vagrant now uses the exact same settings as a staging server, without any extra work copying a bunch of junk around.

It's a little tough to read these changes in the diff, I do suggest grabbing a copy of these files in magfest/production-config and just browsing them there.

binary1230 replied:
ok one other thing to actually answer your original question:

the files don't override each other.  individual settings in the files override each other.

if the hierarchy was:
common.yaml
production.yaml
event-classic.yaml

and you had the following stuff in each

common.yaml
```
dbhost='localhost'

send_emails='False'
dev_box='True'

1day_badge_price='30'
```

production.yaml
```
send_emails='True'
dev_box='False'
```

event-classic.yaml
```
1day_badge_price='50'
```

then what you end up with at the end is a merging of all those settings, like this:
```
dbhost='localhost'
send_emails='True'
dev_box='False'
1day_badge_price='50'
```

The take-away here is each level of the hierarchy only has to define what it needs / what it needs to override.

kitsuta replied:
This all looks pretty good, but keep in mind that we'll want to override some settings in the staging server that don't qualify as secret. Like if we want to test dealer registration before it opens, we need to override those dates. Because of that, I'm still not sure it makes a ton of sense to put the environment YAML above the event YAML.

binary1230 replied:
Ah though that's doable in two ways: ovverride that setting in just stage.uber.magfest.org.yaml 

Or, add a new file or re order sure

binary1230 replied:
Note1: So hey, in fact we did need to add event specific + production specific items so here's the new final hierarchy:

```
  - "nodes/external/secret/%{::fqdn}"
  - "nodes/external/secret/event-%{::event_name}-%{::environment}"
  - "nodes/external/secret/event-%{::event_name}"
  - "nodes/external/secret/%{::environment}"
  - "nodes/external/secret/common"
  - "nodes/external/%{::fqdn}"
  - "nodes/vagrant-%{::event_name}-%{::is_vagrant}"
  - "nodes/vagrant-%{::is_vagrant}"
  - "nodes/event-%{::event_name}-%{::environment}"
  - "nodes/event-%{::event_name}"
  - "nodes/%{::environment}"
  - "nodes/common"
  # things below nodes are included in this repo and visible to any event running uber
  - "vagrant-%{::is_vagrant}"
  - "%{::environment}"
  - "common"
```

For stripe keys which need to be set on a per-environment per-event_name setting, we have this line:
```
  - "nodes/external/secret/event-%{::event_name}-%{::environment}"
```

That means a file might exist called
event-classic-production.yaml

That would have the magclassic production-specific stripe keys in it.

binary1230 replied:
Note2: overriding stuff like reg-open dates for specific staging servers:

We could put that in either:

```nodes/event-classic-development.yaml```

Or we could also create:

```nodes/external/staging2.uber.magfest.org```

In those files, we'd add this line:

```yaml
uber::config::prereg_open: '07-01-15'
```

kitsuta commented on INSTALL-vagrant.md:
````diff
@@ -49,11 +49,17 @@ vagrant up
 ```
 
-5) once in via SSH, you can now start the automated installer:
+5) once in via SSH, you can now start the automated installer.
````
This is a minor nitpick, but if someone is reading this tutorial they are pretty likely to run the command you give them _before_ seeing your alternative below. Maybe just add "optional_event_name" to the end of the `./setup_vagrant_control_server.sh` command.

binary1230 replied:
true, yea

binary1230 replied:
fix'd

binary1230 committed:
update real installation docs

binary1230 committed:
fix init_node.sh params passed to fabric

binary1230 committed:
Merge branch 'big_puppet_refactor' of https://github.com/magfest/ubersystem-deploy into big_puppet_refactor

binary1230 committed:
fix config settings set incorrectly
- they needed uber::config in front of them
remove more magfest-specific defaults from config.pp

binary1230 committed:
Merge remote-tracking branch 'origin/big_puppet_refactor' into big_puppet_refactor

binary1230 committed:
Merge remote-tracking branch 'origin/master' into big_puppet_refactor

Conflicts:
	puppet/hiera/vagrant-1.yaml

binary1230 committed:
add event-%{::event_name}-%{::environment} yaml file to hierarchy
- we'll need this for stripe keys which are environment-specific and event-specific, maybe other stuff

binary1230 committed:
add secrets common.yaml to hierarchy

binary1230 committed:
don't exclude external/ nodes directory in deploy's rsync

unfortunately, due to the new hierarchy, we can't know which hiera files are needed ahead of time, so we can't exclude other node's secret data.

really, we should be using a puppet master which is able to do exactly this, instead of rolling this functionality ourselves, but, that's outside the scope of this refactor

binary1230 committed:
create facts.d dir before use

binary1230 committed:
Merge branch 'big_puppet_refactor' of https://github.com/magfest/ubersystem-deploy into big_puppet_refactor

binary1230 committed:
Update INSTALL-vagrant.md

make this clearer, per @kitsuta's suggestion

binary1230 committed:
print more facts for print_server_info()

binary1230 committed:
no longer ignore stripe/aws keys during compares

binary1230 committed:
Merge branch 'big_puppet_refactor' of https://github.com/magfest/ubersystem-deploy into big_puppet_refactor

binary1230 commented:
re-reviewed, internal +1

binary1230 committed:
ingore more files

binary1230 committed:
remote postgres backup functionality

binary1230 committed:
Merge remote-tracking branch 'origin/big_puppet_refactor' into big_puppet_refactor

binary1230 committed:
fix postgres backup

binary1230 committed:
fix postgres backup creation of database dir if it already exists

binary1230 committed:
switch home dirs correctly when doing backups

binary1230 committed:
change DB name to reflect default setting

binary1230 committed:
backups dir owned by postgres, add additional permissions

binary1230 committed:
fix backups dir permissions
